icon.c

```diff
@@ -0,0 +1,69 @@
+#include <stdio.h>
+#include <assert.h>
+#include <cairo/cairo.h>
+#include <gdk-pixbuf/gdk-pixbuf.h>
+#include <gdk/gdk.h>
+
+#include "icon.h"
+
+GdkPixbuf *load_image(const char *path) {
+	if (strlen(path) == 0) {
+		return NULL;
+	}
+	GError *err = NULL;
+	GdkPixbuf *pixbuf = gdk_pixbuf_new_from_file(path, &err);
+	if (!pixbuf) {
+		fprintf(stderr, "Failed to load icon (%s)\n", err->message);
+		return NULL;
+	}
+	return pixbuf;
+}
+
+struct mako_icon get_icon(const char *path, double max_size) {
+	struct mako_icon icon;
+	icon.image = load_image(path);
+	if (icon.image == NULL)
+		return icon;
+
+	icon.image_width = gdk_pixbuf_get_width(icon.image);
+	icon.image_height = gdk_pixbuf_get_height(icon.image);
+
+	if (icon.image_width > icon.image_height) {
+		// Width limits size
+		if (icon.image_width > max_size) {
+			icon.width = max_size;
+			icon.scale = icon.width/icon.image_width;
+			icon.height = icon.image_height*icon.scale;
+		} else {
+			icon.width = icon.image_width;
+			icon.height = icon.image_height;
+		}
+	} else {
+		// Height limits size
+		if (icon.image_height > max_size) {
+			icon.height = max_size;
+			icon.scale = icon.height/icon.image_height;
+			icon.width = icon.image_width*icon.scale;
+		} else {
+			icon.width = icon.image_width;
+			icon.height = icon.image_height;
+		}
+	}
+
+	return icon;
+}
+
+void draw_icon(cairo_t *cairo, struct mako_icon icon,
+		double xpos, double ypos, double scale) {
+	cairo_save(cairo);
+	cairo_scale(cairo, icon.scale * scale, icon.scale * scale);
+	gdk_cairo_set_source_pixbuf(cairo, icon.image, xpos / icon.scale, ypos / icon.scale);
```

Comment:
I wonder if it would be more efficient to create a new cairo surface and draw the icon into it when loading it, rather than waiting until we draw it to convert from the pixbuf? I don't have strong feelings one way or the other about this though.

Reply:
It might be a good idea to not only make it into a cairo surface, but also to scale it immediately as well (especially if we move loading to `handle_notify`). **Pro:** doesn't store the full resolution icon for long. **Con:** in the odd case of the user changing dpi or `max-icon-size` at runtime an active notification might have a too low res icon.